### ISSUE TITLE
[HOLD for payment 2024-06-05] [$250] Chat - Text with attachment message can be edited when the attachment is uploading

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#40400]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.63-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:
1. Go to staging.new.expensify.com
2. Go to chat.
3. Copy an image (from any source).
4. Enter text in the composer.
5. Paste the image via CMD+V.
6. Send the attachment together with the text.
7. While the attachment is uploading, right click on the message > Edit comment.
8. Edit the comment and save it.

## Expected Result:
In Step 7, app should prevent comment to be editable when the attachment is uploading.

## Actual Result:
In Step 7, app allows editing when the attachment with text is uploading.
As a result, the attachment is not uploaded and becomes "Uploading attachment..." when editing and saving the comment (Step 8).

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/fae7e4c0-8b4e-441c-b397-bc06c7354959

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~016d10a063fea0e0dc</li>
        <li>Upwork Job ID: 1780928057418211328</li>
        <li>Last Price Increase: 2024-04-25</li>
<li>Automatic offers: </li>
<ul>
<li>FitseTLT | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @MitchExpensify</details>

**Comment by izarutskaya** [comment]
@MitchExpensify I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors.

**Comment by izarutskaya** [comment]
We think this issue might be related to the #collect project.

**Comment by izarutskaya** [comment]
Production


https://github.com/Expensify/App/assets/115492554/b0573492-363a-4422-9ce4-1efde964df85

**Comment by shahinyan11** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Chat - Text with attachment message can be edited when the attachment is uploading
### What is the root cause of that problem?
There is not any implementation for this
### What changes do you think we should make in order to solve the problem?
Update this [shouldShow](https://github.com/Expensify/App/blob/3ec8f2f9473ee1ccee7b8fc74b60cf608ee590f5/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L223C9-L224C149) function like below
```
shouldShow: (type, reportAction, isArchivedRoom, betas, menuTarget, isChronosReport) =>{
    const isAttachment = ReportActionsUtils.isReportActionAttachment(reportAction);
    const messageHtml = getActionHtml(reportAction);
    if(isAttachment && messageHtml === CONST.ATTACHMENT_UPLOADING_MESSAGE_HTML){
        return false
    }
    return type === CONST.CONTEXT_MENU_TYPES.REPORT_ACTION && ReportUtils.canEditReportAction(reportAction) && !isArchivedRoom && !isChronosReport
},
```

### What alternative solutions did you explore? (Optional)

**Comment by Nodebrute** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Text with attachment message can be edited when the attachment is uploading #
### What is the root cause of that problem?
Here, isAttachment will be false. That's why we can see the edit icon for comments that have text and attachments.
https://github.com/Expensify/App/blob/3ec8f2f9473ee1ccee7b8fc74b60cf608ee590f5/src/libs/ReportActionsUtils.ts#L845
### What changes do you think we should make in order to solve the problem?

1. We can set isAttachment to true in reportAction for cases like these.
2. Or we can simply check if reportAction.attachmentInfo is true. We can do something like this
 ```
reportAction?.attachmentInfo !== undefined
```
https://github.com/Expensify/App/blob/3ec8f2f9473ee1ccee7b8fc74b60cf608ee590f5/src/libs/ReportActionsUtils.ts#L845
3. We can create a new utility function hasAttachment for cases like these.
### What alternative solutions did you explore? (Optional)

**Comment by shahinyan11** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Chat - Text with attachment message can be edited when the attachment is uploading
### What is the root cause of that problem?
[Here](https://github.com/Expensify/App/blob/3ec8f2f9473ee1ccee7b8fc74b60cf608ee590f5/src/libs/ReportActionsUtils.ts#L844C1-L850C6) the queue of `if` block are wrong
### What changes do you think we should make in order to solve the problem?
Move [this](https://github.com/Expensify/App/blob/3ec8f2f9473ee1ccee7b8fc74b60cf608ee590f5/src/libs/ReportActionsUtils.ts#L848C5-L850C6) if block above [this](https://github.com/Expensify/App/blob/3ec8f2f9473ee1ccee7b8fc74b60cf608ee590f5/src/libs/ReportActionsUtils.ts#L844C5-L846C6)

### What alternative solutions did you explore? (Optional)

**Comment by mollfpr** [comment]
@Nodebrute @shahinyan11 Did you guys find the offending PR?

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Text with attachment message can be edited when the attachment is uploading

### What is the root cause of that problem?
It is a regression from https://github.com/Expensify/App/pull/39007 which modifies optimistic data for text+attachment but doesn't align with what BE returns which is `isAttachment` is not defined
### What changes do you think we should make in order to solve the problem?

Change this 
https://github.com/Expensify/App/blob/3ec8f2f9473ee1ccee7b8fc74b60cf608ee590f5/src/libs/ReportUtils.ts#L3223-L3224
to
```
    const isAttachment = text && file !== undefined ? undefined : !text && file !== undefined;

```
### What alternative solutions did you explore? (Optional)

**Comment by mollfpr** [comment]
>We can set isAttachment to true in reportAction for cases like these.

@Nodebrute `isAttachment` also true after the message is sent?

@shahinyan11 I try your solution but it's not working.

@FitseTLT I think your solution will block the user after the message is sent.

**Comment by mollfpr** [comment]
>We have to get the message HTML and check for messageHtml !== CONST.ATTACHMENT_UPLOADING_MESSAGE_HTML.

@ShridharGoel  Where is this check implemented?

**Comment by FitseTLT** [comment]
@mollfpr my solution will make it perfectly aligned with what the BE returns and will solve the current issue from its root cause but if we want to edit text+attachment kind of messages, that will be another issue and we have to confirm if that is desired and we can modify the current canEdit logic

**Comment by MariaHCD** [comment]
Thanks, @mollfpr! I see the conversation was started here: https://expensify.slack.com/archives/C02NK2DQWUX/p1713444816693779

**Comment by shahinyan11** [comment]
@mollfpr I am sorry . I mistake the  `()` in my alternative solution. There should be  
```
return (reportAction?.isAttachment || !!reportAction?.attachmentInfo) ?? false;
```

**Comment by MariaHCD** [comment]
After the discussion [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1713444816693779), this isn't a deploy blocker. Let's go with the solution of disabling editing while the file is uploading.

**Comment by MariaHCD** [comment]
We've landed on the expected behavior. Waiting on @mollfpr to review all the proposals so far.

**Comment by mollfpr** [comment]
@FitseTLT I tried your solution and we have a couple of issues.


https://github.com/Expensify/App/assets/25520267/85bb4255-128d-434a-9e93-3b05e13c9c07


1. The action to edit the comment is not showing right after the upload finishes, but instead waiting for the image to be downloaded.
2. Can't edit the message sent while offline

---

@shahinyan11 Same issue, can't edit the message while offline. Also, what's the reason for this change?

`return (reportAction?.isAttachment || !!reportAction?.attachmentInfo) ?? false;`

**Comment by shahinyan11** [comment]
> @shahinyan11 Same issue, can't edit the message while offline. Also, what's the reason for this change?
> 
> `return (reportAction?.isAttachment || !!reportAction?.attachmentInfo) ?? false;`

@mollfpr Regarding the inability to edit a message offline, this is the current behavior of the application. My proposed changes do not affect to this behavior.
The reason of  above change - In `reportAction` containing text and attachment `isAttachment` is `false`  but `attachmentInfo` is exists. And current [implementation](https://github.com/Expensify/App/blob/3ec8f2f9473ee1ccee7b8fc74b60cf608ee590f5/src/libs/ReportActionsUtils.ts#L845) always returns `false` in this case because `false ?? true ?? false` returns `false`.  But I think if the `attachmentInfo` exists the condition should return `true`

**Comment by mollfpr** [comment]
>Regarding the inability to edit a message offline, this is the current behavior of the application.

@shahinyan11 Currently in production we can edit a message offline, and I never heard that we can't. Could you elaborate more?

https://github.com/Expensify/App/assets/25520267/9c5ea63b-2234-4248-b55b-3bdec49cb50c

**Comment by shahinyan11** [comment]
@mollfpr I just tested lust version on local and there is not `edit` action  when I send message offline

**Comment by FitseTLT** [comment]
>Can't edit the message sent while offline

Oh yeah @mollfpr it's because attachmentInfo is empty object by default so I had to use `isEmptyObject` check instead so updated accordingly.
```
 !isReportMessageAttachment(reportAction?.message?.[0]) &&
            !(!isEmptyObject(reportAction.attachmentInfo) && reportAction.isOptimisticAction) &&
```

**Comment by mollfpr** [comment]
@shahinyan11 What version do you use?

**Comment by shahinyan11** [comment]
> @shahinyan11 What version do you use?

@mollfpr Ah sorry . I was wrong.

**Comment by FitseTLT** [comment]
>Can't edit the message sent while offline

Oh yeah @mollfpr it's because attachmentInfo is empty object by default so I had to use `isEmptyObject` check instead so updated accordingly.
```
 !isReportMessageAttachment(reportAction?.message?.[0]) &&
            !(!isEmptyObject(reportAction.attachmentInfo) && reportAction.isOptimisticAction) &&
```

>The action to edit the comment is not showing right after the upload finishes, but instead waiting for the image to be downloaded.

Its not like the logic is depending on the image download status but it takes a little more time to reset `isOptimisticAction`  on success of the request. But `isOptimisticAction` is by far the right and reliable way of checking for if the uploading is finished but if we depend on the text being `UPLOADING...` or sth if the user inputs that same text we might end up assuming it is still uploading forever.

**Comment by mollfpr** [comment]
@ShridharGoel Yes, I still reviewing your proposal and the others.

**Comment by shahinyan11** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/40400#issuecomment-2063503608). Fix issue Can't edit the message sent while offline

**Comment by MitchExpensify** [comment]
Friendly bump @mollfpr

**Comment by mollfpr** [comment]
@ShridharGoel I test your solution but it's still showing the edit action on the image uploading.

In case my changes might be wrong, here's the diff.

```diff
diff --git a/src/libs/ReportActionsUtils.ts b/src/libs/ReportActionsUtils.ts
index 71d66a5d2a..7b4e01caff 100644
--- a/src/libs/ReportActionsUtils.ts
+++ b/src/libs/ReportActionsUtils.ts
@@ -844,7 +844,7 @@ function isReportActionAttachment(reportAction: OnyxEntry<ReportAction>): boolea
     const message = reportAction?.message?.[0];

     if (reportAction && ('isAttachment' in reportAction || 'attachmentInfo' in reportAction)) {
-        return reportAction?.isAttachment ?? !!reportAction?.attachmentInfo ?? false;
+        return (reportAction?.isAttachment ?? !!reportAction?.attachmentInfo) ?? false;
     }

     if (message) {
diff --git a/src/libs/ReportUtils.ts b/src/libs/ReportUtils.ts
index a0da34416a..a62f0298b8 100644
--- a/src/libs/ReportUtils.ts
+++ b/src/libs/ReportUtils.ts
@@ -2592,6 +2592,12 @@ function canEditFieldOfMoneyRequest(reportAction: OnyxEntry<ReportAction>, field
     return true;
 }

+function getActionHtml(reportAction: OnyxEntry<ReportAction>): string {
+    const message = reportAction?.message?.at(-1) ?? null;
+    return message?.html ?? '';
+}
+
+
 /**
  * Can only edit if:
  *
@@ -2603,11 +2609,18 @@ function canEditFieldOfMoneyRequest(reportAction: OnyxEntry<ReportAction>, field
 function canEditReportAction(reportAction: OnyxEntry<ReportAction>): boolean {
     const isCommentOrIOU = reportAction?.actionName === CONST.REPORT.ACTIONS.TYPE.ADD_COMMENT || reportAction?.actionName === CONST.REPORT.ACTIONS.TYPE.IOU;

+    const isAttachmentInUploadingState = () => {
+        const messageHtml = getActionHtml(reportAction);
+        return messageHtml.includes(CONST.ATTACHMENT_UPLOADING_MESSAGE_HTML);
+    }
+
+    const shouldShowEditForAttachment = ReportActionsUtils.isReportActionAttachment(reportAction) && !isAttachmentInUploadingState()
+
     return Boolean(
         reportAction?.actorAccountID === currentUserAccountID &&
             isCommentOrIOU &&
             canEditMoneyRequest(reportAction) && // Returns true for non-IOU actions
-            !ReportActionsUtils.isReportActionAttachment(reportAction) &&
+            (!ReportActionsUtils.isReportActionAttachment(reportAction) || shouldShowEditForAttachment) && // Either it's not an attachment, or if it is an attachment then it shouldn't be in uploading state
             !ReportActionsUtils.isDeletedAction(reportAction) &&
             !ReportActionsUtils.isCreatedTaskReportAction(reportAction) &&
             reportAction?.pendingAction !== CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE,
```

**Comment by mollfpr** [comment]
From all the solutions I tried, the closest to the expected result is @FitseTLT alternative [proposal](https://github.com/Expensify/App/issues/40400#issuecomment-2063741582), and it's a simple fix for this issue.

🎀 👀 🎀 C+ reviewed!

**Comment by mollfpr** [comment]
@ShridharGoel Yes, I did.

**Comment by mollfpr** [comment]
>Also, for checking the uploading state in existing flows, we check the HTML and don't use isOptimisticAction.

Using the HTML will have flaws, for example, if I send a text message that contains `Uploading attachment...` the edit action will not show.

**Comment by shahinyan11** [comment]
> @shahinyan11 I also tried your latest update, but now it's not working.

@mollfpr I will check today and give updates

**Comment by MariaHCD** [comment]
@mollfpr looks like we're still evaluating the proposals here?

**Comment by MitchExpensify** [comment]
Bump @mollfpr

**Comment by MariaHCD** [comment]
Thanks, @mollfpr! @FitseTLT's alternative proposal makes sense

**Comment by mollfpr** [comment]
@FitseTLT Let me know when the PR ready, thanks!

**Comment by MariaHCD** [comment]
Not overdue; we're waiting on a PR from @FitseTLT

**Comment by MitchExpensify** [comment]
Reminder set to pay, thoughts on BZ steps @mollfpr ?

**Comment by MitchExpensify** [comment]
*Payment summary:*

- $250 @mollfpr requires payment through NewDot Manual Requests
- $250 @FitseTLT requires payment [automatic offer](https://www.upwork.com/nx/wm/offer/0) (Contributor)

**Comment by MitchExpensify** [comment]
Paid and contract ended @FitseTLT

**Comment by mollfpr** [comment]
Sorry for the delay 🙏

>[@mollfpr] The PR that introduced the bug has been identified. Link to the PR:
 [@mollfpr] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

No offending PR was found.

> [@mollfpr] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

This is an improvement for a new feature, so the regression step should be enough.

> [@mollfpr] Determine if we should create a regression test for this bug.
 [@mollfpr] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

1. Go to staging.new.expensify.com
2. Go to chat.
3. Copy an image (from any source).
4. Enter text in the composer.
5. Paste the image via CMD+V.
6. Send the attachment together with the text.
7. While the attachment is uploading, right-click on the message
8. Verify Edit comment is not included in the context menu list
9. Wait until the attachment is uploaded and open the context menu on the message
10. Verify Edit comment menu exists and press on it
11. Check that now the text+attachment message can be edited
12. 👍 or 👎

**Comment by MitchExpensify** [comment]
Thank you!

**Comment by JmillsExpensify** [comment]
$250 approved for @mollfpr
